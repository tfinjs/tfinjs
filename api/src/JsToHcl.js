import assert from 'assert';
import requiredParam from './statics/requiredParam';

/* an arrayLike is e.g.

wef {
  filter {
    hi = 123
  }
  filter {
    ho = 321
  }
}

which is represented by

wef: {
  filter: [
    {
      hi: 123,
    },
    {
      ho: 321,
    }
  ],
},

*/

const isPrimitive = (val) =>
  ['string', 'boolean', 'number'].includes(typeof val);

const parsePrimitive = (val) => {
  switch (typeof val) {
    case 'string':
      return val.match(/"/) ? `<<EOF\n${val}\nEOF` : `"${val}"`;
    case 'number':
      return val;
    case 'boolean':
      return val ? 'true' : 'false';
    default:
      break;
  }
  return val;
};

class JsToHclUtils {
  /**
   * Converts a javascript object or array to HCL
   *
   * @param {object|array} js
   * @returns hcl
   * @memberof JsToHcl
   */
  stringify(js) {
    const hcl = this.parse(js);
    return hcl.replace(/^\{/, '').replace(/\}$/, '');
  }

  /**
   * converts any javascript type to hcl
   *
   * @returns hcl
   * @memberof JsToHcl
   */
  parse = (value = requiredParam('value')) => {
    assert(
      !['undefined', 'symbol'].includes(typeof value)
        && value !== null
        && !Number.isNaN(value),
      'Value cannot be null, undefined, NaN, symbol',
    );
    if (isPrimitive(value)) {
      return parsePrimitive(value);
    }
    if (Array.isArray(value)) {
      return `[
        ${value.map(this.parse).join(',\n')}
      ]`;
    }

    if (value instanceof JsToHclUtils) {
      return value.stringify();
    }

    assert(
      typeof value !== 'function',
      'The only function you can provide is an instance of a Provisioner',
    );

    return `{
      ${Object.entries(value)
      .map(this.createHclKeyVal)
      .sort()
      .join('\n')}
    }`;
  };

  /**
   * Creates key value pair in hcl
   *
   * @param {array} params
   * @param {string} params.0 - key
   * @param {string} params.1 - value
   * @returns hcl
   */
  createHclKeyVal = ([
    key = requiredParam('key'),
    value = requiredParam('value'),
  ]) => {
    if (this.isArrayLike(value)) {
      return this.parseArrayLike(key, value);
    }
    const parsedValue = this.parse(value);
    if (value instanceof JsToHclUtils) {
      return parsedValue;
    }
    return `${key} = ${parsedValue}`;
  };

  isArrayLike = (value) =>
    Array.isArray(value) && value.every((val) => typeof val === 'object');

  parseArrayLike = (key, value) =>
    value.map((val) => `${key} ${this.parse(val)}`);

  getBody(js) {
    return this.parse(js)
      .replace(/^[{[]/, '')
      .replace(/[}\]]$/, '');
  }
}

export class ArrayLike extends JsToHclUtils {
  constructor(definition, body) {
    super();
    assert(typeof definition === 'string', 'definition must be a string');
    assert(typeof body === 'object', 'body must be an object');
    this.definition = definition;
    this.body = body;
  }

  stringify() {
    return `${this.definition} { ${this.getBody(this.body)} } `;
  }
}

export class Provisioner extends ArrayLike {
  constructor(type, body) {
    assert(typeof type === 'string', 'type must be a string');
    assert(typeof body === 'object', 'body must be an object');
    super(`provisioner "${type}"`, body);
  }
}

/**
 * Converts JavaScript to HCL
 *
 * @class JsToHcl
 */
class JsToHcl extends JsToHclUtils {}

export default JsToHcl;
