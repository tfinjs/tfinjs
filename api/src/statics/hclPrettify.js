import { execFile } from 'child_process';

const hclPrettify = (hcl) => new Promise((resolve, reject) => {
  const cp = execFile('terraform', ['fmt', '-'], {
    cwd: process.cwd(),
    stdio: ['pipe', 'pipe', 'inherit'],
  });
  let string = '';
  cp.stdout.on('data', (part) => {
    string += part;
  });

  cp.stdout.on('end', () => {
    resolve(string);
  });

  cp.stdout.on('error', (err) => {
    reject(err);
  });

  cp.stdin.write(hcl);
  cp.stdin.end();
});

export default hclPrettify;
