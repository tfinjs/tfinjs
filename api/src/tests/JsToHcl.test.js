import { resolve } from 'path';
import mkdrip from 'mkdirp';
import { writeFileSync, readFileSync } from 'fs';
import JsToHcl, { Provisioner } from '../JsToHcl';
import hclPrettify from '../statics/hclPrettify';

/* eslint-env jest */

const testObject = {
  glossary: {
    title: 'example glossary',
    GlossDiv: {
      title: 'S',
      value: 729,
      repeat: true,
      GlossList: {
        GlossEntry: {
          ID: 'SGML',
          SortAs: 'SGML',
          GlossTerm: 'Standard Generalized Markup Language',
          Acronym: 'SGML',
          Abbrev: 'ISO 8879:1986',
          GlossDef: {
            para:
              'A meta-markup language, used to create markup languages such as DocBook.',
            GlossSeeAlso: ['GML', 'XML'],
          },
          GlossSee: 'markup',
        },
      },
    },
  },
};

test('JsToHcl default', async () => {
  const jsToHcl = new JsToHcl();
  const result = jsToHcl.stringify(testObject);

  const prettyResult = await hclPrettify(result);

  expect(prettyResult).toMatchSnapshot();
});

/* add provisioner */

test('JsToHcl with provisioner', async () => {

  const js = {
    'provisioner "local-exec"': [
      {
        command: 'echo first',
      },
      {
        command: 'echo second',
      },
    ],
  };
  const jsToHcl = new JsToHcl();
  const result = jsToHcl.stringify(js);

  const prettyResult = await hclPrettify(result);

  expect(prettyResult).toMatchSnapshot();
});
