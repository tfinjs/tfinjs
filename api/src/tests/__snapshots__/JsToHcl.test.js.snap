// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JsToHcl default 1`] = `
"glossary = {
  GlossDiv = {
    GlossList = {
      GlossEntry = {
        Abbrev  = \\"ISO 8879:1986\\"
        Acronym = \\"SGML\\"

        GlossDef = {
          GlossSeeAlso = [
            \\"GML\\",
            \\"XML\\",
          ]

          para = \\"A meta-markup language, used to create markup languages such as DocBook.\\"
        }

        GlossSee  = \\"markup\\"
        GlossTerm = \\"Standard Generalized Markup Language\\"
        ID        = \\"SGML\\"
        SortAs    = \\"SGML\\"
      }
    }

    repeat = true
    title  = \\"S\\"
    value  = 729
  }

  title = \\"example glossary\\"
}
"
`;

exports[`JsToHcl with provisioner 1`] = `
"provisioner \\"local-exec\\" {
  command = \\"echo first\\"
}

provisioner \\"local-exec\\" {
  command = \\"echo second\\"
}
"
`;
